src/shared/cloudformation/templateSymbolResolver.ts

```diff
@@ -27,34 +31,70 @@ export interface TemplateSymbolProvider {
+     *
+     * - Function (no Api): `AWS::Serverless::Function` _without_ `Events.Type=Api`.
+     * - Implicit Api: `AWS::Serverless::Function` with `Events.Type=Api`.
+     * - Explicit Api: `AWS::Serverless::Api`.
      */
-    public async getFunctionResources(): Promise<TemplateFunctionResource[]> {
-        const symbols = await this.symbolProvider.getSymbols(this.document)
-
-        const resources = _(symbols).find({ name: 'Resources', kind: vscode.SymbolKind.Module })?.children ?? []
-
-        return _(resources)
+    public async getResourcesOfKind(kind: 'function' | 'api'): Promise<TemplateFunctionResource[]> {
+        const allSymbols = await this.symbolProvider.getSymbols(this.document)
+        const wantSymbol = kind === 'function' ? 'Resources' : 'Events'
+        let symbols
+        if (kind === 'function') {
+            // Only want top level (todo: is this actually true?)
+            symbols = _(allSymbols).find({ name: wantSymbol, kind: vscode.SymbolKind.Module })?.children ?? []
+        } else {
+            symbols = this.findDescendants(allSymbols, wantSymbol, vscode.SymbolKind.Module)
+        }
+        return _(symbols)
             .filter({ kind: vscode.SymbolKind.Module })
-            .filter(resource => this.isFunctionResource(resource))
-            .map(resource => ({ name: resource.name, range: resource.range }))
+            .filter(r => this.isKind(kind, r))
+            .map(r => ({ name: r.name, range: r.range, kind: kind }))
             .value()
     }
 
-    private isFunctionResource(symbol: vscode.DocumentSymbol) {
-        const typeSymbol = _(symbol.children).find({ name: 'Type', kind: vscode.SymbolKind.String })
+    /**
+     * Searches the tree and returns all matching nodes.
+     */
+    private findDescendants(
+        symbols: vscode.DocumentSymbol[],
+        name: string,
+        kind: vscode.SymbolKind
+    ): vscode.DocumentSymbol[] {
+        const found = symbols.filter(v => v.name === name && v.kind === kind)
+        for (let s of symbols) {
+            found.push(...this.findDescendants(s.children, name, kind))
+        }
+        return found
+    }
 
+    private isKind(kind: 'function' | 'api', symbol: vscode.DocumentSymbol) {
+        // Example:
+        // ------------------------------------
+        // Resources:
+        //   ImplicitApiFunction:
+        //     Type: AWS::Serverless::Function
+        //     Properties:
+        //       ...
+        //       Events:             \
+        //         GetHtml:           \
+        //           Type: Api         \ presence of this means
+        //           Properties:       / API instead of FUNCTION
+        //             Path: /        /
+        //             Method: get   /
+        const typeSymbol = this.findDescendants(symbol.children, 'Type', vscode.SymbolKind.String)[0]
+        const wantType = kind === 'function' ? CloudFormation.SERVERLESS_FUNCTION_TYPE : 'Api'
```

Comment:
Would this tromp over `AWS::Lambda::Function` and is there a difference? Also should this be `: 'AWS::Serverless::Api`?

Reply:
good catch, fixed. Reworked the tests in 795525cf385c, and the test fails correctly before fixing this.

Reply:
Actually this was correct as it was. In template.yaml the declaration is `Type: Api` which is a descendant of a `Type: AWS::Serverless::Function` node. Those are the nodes that `isKind()` searches for.

Reply:
OK that makes sense. Can you change the doc comment to specifically highlight what `function` and `api` are looking for? 

If this function starts looking for other types, we may want to break it out into separate functions.